csharp/src/Ice/Communicator.cs

```diff
@@ -1364,5 +1273,106 @@ private void AddAllAdminFacets()
 
             return null;
         }
+
+        private async Task PerformDisposeAsync()
+        {
+            lock (_mutex)
+            {
+                Debug.Assert(_state < StateDestroyInProgress);
+                _state = StateDestroyInProgress;
+            }
+
+            // Cancel operations that are waiting and using the communicator's cancellation token
+            _cancellationTokenSource.Cancel();
+
+            // Shutdown and destroy all the incoming and outgoing Ice connections and wait for the connections to be
+            // finished.
+            _ = ShutdownAsync();
+
+            _outgoingConnectionFactory?.DestroyAsync();
+
+            await WaitForShutdownAsync().ConfigureAwait(false);
+            _shutdownTaskSemaphore?.Dispose();
+
+            // At this point adapters are immutable
+            foreach (ObjectAdapter adapter in _adapters)
+            {
+                await adapter.DisposeAsync().ConfigureAwait(false);
+            }
+
+            // _adminAdapter is disposed above when iterating over all adapters, we call DisposeAsync here to avoid the
+            // compiler warning about disposable field not being dispose.
+            if (_adminAdapter != null)
+            {
+                await _adminAdapter.DisposeAsync().ConfigureAwait(false);
+            }
+
+            if (_outgoingConnectionFactory != null)
+            {
+                await _outgoingConnectionFactory.DestroyAsync().ConfigureAwait(false);
+            }
+
+            Observer?.SetObserverUpdater(null);
+
+            if (Logger is LoggerAdminLogger adminLogger)
+            {
+                await adminLogger.DisposeAsync().ConfigureAwait(false);
+            }
+
+            if (GetPropertyAsBool("Ice.Warn.UnusedProperties") ?? false)
+            {
+                List<string> unusedProperties = GetUnusedProperties();
+                if (unusedProperties.Count != 0)
+                {
+                    var message = new StringBuilder("The following properties were set but never read:");
+                    foreach (string s in unusedProperties)
+                    {
+                        message.Append("\n    ");
+                        message.Append(s);
+                    }
+                    Logger.Warning(message.ToString());
+                }
+            }
+
+            // Destroy last so that a Logger plugin can receive all log/traces before its destruction.
+            List<(string Name, IPlugin Plugin)> plugins;
+            lock (_mutex)
+            {
+                plugins = new List<(string Name, IPlugin Plugin)>(_plugins);
+            }
```

Comment:
Is it really needed to copy the _plugins list here? Plugins can still be added/removed while the communicator is being disposed?

Reply:
seems not, maybe we should just use `Enumerable.Reverse`
```
foreach ((string name, IPlugin plugin) in Enumerable.Reverse(_plugins))
{
...
}
```

Reply:
Sounds fine.